File: src/export_csv.py
#
# Copyright 2020 Lukas Schmelzeisen
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import csv
from operator import attrgetter
from pathlib import Path
from shutil import rmtree

import nasty  # type: ignore
from tqdm import tqdm

crawl_dir = Path("twitter-crawl")
csv_dir = crawl_dir.parent / (crawl_dir.name + ".csv")
if csv_dir.exists():
    rmtree(csv_dir)

twitter_crawl = nasty.BatchResults(crawl_dir)
for entry in tqdm(twitter_crawl, desc="Twitter-Crawl-Entries"):
    csv_fields = [
        "url",
        "created_at",
        "reply_count",
        "retweet_count",
        "favorite_count",
        "full_text",
    ]
    tweets = sorted(twitter_crawl.tweets(entry), key=attrgetter("created_at"))

    file = (
        csv_dir
        / entry.request.filter.name.lower()
        / entry.request.lang
        / "{}-{}.csv".format(entry.request.query.replace(" ", "-"), entry.request.since)
    )
    Path.mkdir(file.parent, exist_ok=True, parents=True)

    with open(file, "w", encoding="UTF-8", newline="") as fin:
        csv_writer = csv.DictWriter(fin, csv_fields)
        csv_writer.writeheader()

        for tweet in tweets:
            tweet_json = tweet.to_json()
            tweet_csv = {
                "url": tweet.url,
                "created_at": tweet.created_at,
                "full_text": tweet.text.replace("\n", " "),
                "is_answer_to_quote": "quoted_status_id_str" in tweet_json,
                "is_reply_to_status": "in_reply_to_status_id_str" in tweet_json,
                "retweet_count": tweet_json["retweet_count"],
                "favorite_count": tweet_json["favorite_count"],
                "reply_count": tweet_json["reply_count"],
                "user_name": tweet.user.name,
                "user_screen_name": tweet.user.screen_name,
                "user_description": tweet_json["user"]["description"],
                "user_followers_count": tweet_json["user"]["followers_count"],
                "user_friends_count": tweet_json["user"]["friends_count"],
                "user_verified": tweet_json["user"]["verified"],
                "user_statuses_count": tweet_json["user"]["statuses_count"],
            }
            csv_writer.writerow({field: tweet_csv[field] for field in csv_fields})

File: src/nasty_analysis/serve/figures/export_csv.js
/*
 * Copyright 2019-2020 Lukas Schmelzeisen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function source_to_csv() {
  var columns = []
  var row = [];
  for (var column in source.data) {
    if (!source.data.hasOwnProperty(column))
      continue;
    columns.unshift(column);
    row.unshift('"' + column + '"');
  }

  var lines = [row.join(",")]
  for (var i = 0; i !== source.get_length(); ++i) {
    var row = [];
    for (var j = 0; j !== columns.length; ++j) {
      var column = columns[j];
      var value = source.data[column][i];
      if (typeof value == "string")
        value = '"' + value.toString().replace(/"/g, '""') + '"';
      else
        value = value.toString();
      row.push(value);
    }
    lines.push(row.join(","));
  }

  return lines.join("\n").concat("\n");
}

var filename = "data.csv";
var filetext = source_to_csv()
var blob = new Blob([filetext], { type: "text/csv; charset=UTF-8" });

if (navigator.msSaveBlob)
  navigator.msSaveBlob(blob, filename);
else {
  const link = document.createElement("a");
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  link.target = "_blank";
  link.style.visibility = "hidden";
  link.dispatchEvent(new MouseEvent("click"));
}
